### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Colin Jaques

### DISCUSSION THREAD

**Discussion by CoJaques** [heig-vd-dai-course/heig-vd-dai-course#73] [Show and tell]
https://github.com/CoJaques/heig-vd-dai-course-java-io

Voici mes résultats sous forme de table générée par chatGpt à l'aide de la sortie de mon programme :

Opération | Taille du fichier | Type de Stream | Temps écoulé (ms) | Temps écoulé (ns)
-- | -- | -- | -- | --
Écriture | 1B | Non-buffered | 0 | 200764
Écriture | 1B | Buffered | 0 | 94348
Lecture | 1B | Non-buffered | 0 | 63920
Lecture | 1B | Buffered | 0 | 164037
Écriture | 1KB | Non-buffered | 1 | 1327557
Écriture | 1KB | Buffered | 0 | 49622
Lecture | 1KB | Non-buffered | 0 | 434599
Lecture | 1KB | Buffered | 0 | 25806
Écriture | 1MB | Non-buffered | 1087 | 1087118884
Écriture | 1MB | Buffered | 3 | 3000466
Lecture | 1MB | Non-buffered | 354 | 354336406
Lecture | 1MB | Buffered | 0 | 917681
Écriture | 10MB | Non-buffered | 10452 | 10452243273
Écriture | 10MB | Buffered | 13 | 13370577
Lecture | 10MB | Non-buffered | 3485 | 3485966684
Lecture | 10MB | Buffered | 8 | 8420376

# Conclusions
## Quel type de stream est le plus efficace pour chaque cas d'utilisation ?
Pour les petits fichiers (1B, 1KB), le stream buffered est généralement plus rapide.
Pour les gros fichiers (1MB, 10MB), le stream buffered est nettement plus rapide.

## Pourquoi est-il plus efficace que les autres types de streams ?
Le buffering réduit le nombre d'opérations d'E/S, ce qui accélère le processus.

## Quelle est la différence entre les données binaires et les données textuelles ?
Les données binaires sont des données brutes, tandis que les données textuelles sont des caractères encodés.

## Qu'est-ce qu'un encodage de caractères ?
C'est un système qui mappe des caractères à des valeurs binaires.

##Pourquoi cette méthodologie est-elle importante ?
Elle permet de comprendre les performances et de choisir le bon type de stream pour chaque cas d'utilisation.

**Comment by ludelafo** [comment]
Merci d'avoir cité la source pour la génération des tables, bonne utilisation de ChatGPT :)

Je rajouterai que cette méthodologie scientifique vous permet de garantir, dans le temps, avec des métriques tangibles, clairs et validées, que les performances de vos applications ne vont pas se dégrader et qu'elles restent adaptées aux besoins actuels.

Quand vous traitez de fichiers textes, n'oubliez pas de spécifier l'encodage (le charset) des fichiers lorsque vous les lisez et que vous les écrivez! 

Un des constructeurs des classes `FileReader` et `FileWriter` permet de spécifier le charset pour être sûr que les caractères soient bien interprétés par votre application: 

- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileReader.html
- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileWriter.html
